### PR TITLE
[ROCm] Triton in XLA for ROCm - ir_emitter_triton related changes.

### PR DESCRIPTION
**Pull request by zoranjovanovic-ns**
Second commit of the series for enabling Triton in XLA for ROCm.

### CODE REVIEW

File: xla/service/gpu/ir_emitter_triton.cc [hunk 1]
```diff
@@ -122,14 +126,21 @@ limitations under the License.
 #include "tsl/platform/path.h"
 #include "tsl/platform/status.h"
 #include "tsl/platform/statusor.h"
+#ifdef TENSORFLOW_USE_ROCM
```
Comment (ddunl) [inline xla/service/gpu/ir_emitter_triton.cc:129]: I think it'd be good if possible to minimize the number of ifdefs, so maybe ifdefs used everywhere at the top, then ```
#ifdef GOOGLE_CUDA
<includes> 
#ifdef TENSORFLOW_USE_ROCM
<includes>
``` or similar.

File: xla/service/gpu/ir_emitter_triton.cc [hunk 2]
```diff
@@ -744,11 +758,23 @@ absl::StatusOr<Value> EmitScope(
...
-    mlir::OpPassManager& pm, const se::CudaComputeCapability& cc,
+    mlir::OpPassManager& pm, const se::GpuComputeCapability& cc,
     const TritonGemmConfig& config,
     mt::nvidia_gpu::ClusterInfo& out_cluster_info) {
-  const int ccAsInt = cc.major * 10 + cc.minor;
+#else
+absl::Status CreateTritonPipeline(
+    mlir::OpPassManager& pm, const se::GpuComputeCapability& cc,
+    const TritonGemmConfig& config) {
+#endif
```
Comment (tdanyluk) [inline xla/service/gpu/ir_emitter_triton.cc:770]: In general I would recommend to reduce the number of ifdefs inside the code, to close to 0.
Could we have some kind of abstraction, like have a 
Something class, and a derived CudaSomething and RocmSomething and just use a runtime if based on whether the se::GpuComputeCapability is cuda or rocm?

**Review by tdanyluk — CHANGES_REQUESTED** [review]

File: xla/service/gpu/ir_emitter_triton.cc (hunk 2 above)
Comment (tdanyluk) [reply xla/service/gpu/ir_emitter_triton.cc:770]: I talked with my colleagues and a class is not really needed.
We could use something like this:

something.h:
```
int something();
```

something_cuda.cc:
```
int something() {
 // implemented with cuda
}
```

something_rocm.cc:
```
int something() {
 // implemented with rocm
}
```
And then build either something_rocm.cc or something_cuda.cc with a select.
So ir_emitter_triton.cc would remain mostly device vendor independent and just include the shared "something.h'.

Comment (zoranjovanovic-ns) [reply xla/service/gpu/ir_emitter_triton.cc:770]: Thank you for the proposal and clarification. I will change the implementation (based on proposal) and update PR.

Comment (tdanyluk) [reply xla/service/gpu/ir_emitter_triton.cc:770]: thank you!

File: xla/service/gpu/ir_emitter_triton.cc [hunk 3]
```diff
@@ -408,12 +418,16 @@ Value EmitElementwise(ImplicitLocOpBuilder& b, absl::string_view libdevice_path,
...
-          inputs[0].getType(), inputs, "libdevice", libdevice_path,
-          ObtainDeviceFunctionName(dev_fn_id.value(),
-                                   hlo.shape().element_type(),
-                                   llvm::Triple("nvptx64-unknown-unknown")),
-          /*pure=*/true);
+        llvm::Triple triple("nvptx64-unknown-unknown");
+        if (std::holds_alternative<se::CudaComputeCapability>
+                  (device_info.gpu_compute_capability())) {
+          triple.setTriple("amdgcn-unknown-unknown");
+        }
```
Comment (tdanyluk) [inline xla/service/gpu/ir_emitter_triton.cc:425]: Is there an error here? It looks like you set an AMD triple when the GPU is NVIDIA.

**Review by tdanyluk — CHANGES_REQUESTED** [review]
Thanks, it looks really good after the refactor.

File: xla/service/gpu/ir_emitter_triton.cc (hunk 3 above)
Comment (zoranjovanovic-ns) [reply xla/service/gpu/ir_emitter_triton.cc:425]: Yes, I am about to push fixed version. Sorry for inconvenience.

File: xla/service/gpu/ir_emitter_triton.cc [hunk 4]
```diff
@@ -148,6 +149,15 @@ using ::mlir::Type;
...
 
+
+std::string GetLibdevicePath(const HloModuleConfig& hlo_config,
+                             const se::DeviceDescription& device_info);
+
+absl::Status CreateTritonPipeline(
+    mlir::OpPassManager& pm, const se::GpuComputeCapability& cc,
+    const TritonGemmConfig& config,
+    mt::nvidia_gpu::ClusterInfo& out_cluster_info);
+
```
Comment (tdanyluk) [inline xla/service/gpu/ir_emitter_triton.cc:160]: It's an interesting choice that these are not in the header.
I mean I get the point that these are kindof private to ir_emitter_triton, but still, we usually put function declarations in headers.
If it's not too difficult, please try to do that.

**Review by tdanyluk — CHANGES_REQUESTED** [review]
Hi. Sorry for getting back to you once more, I just noticed some things.

File: xla/service/gpu/ir_emitter_triton.cc (hunk 4 above)
Comment (zoranjovanovic-ns) [reply xla/service/gpu/ir_emitter_triton.cc:160]: OK, I will made changes that you have suggested.

**Review by tdanyluk — APPROVED** [review]
Thanks again, looks great!

File: xla/service/gpu/ir_emitter_triton_cuda.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+std::string GetLibdevicePath(const HloModuleConfig& hlo_config,
+                             const se::DeviceDescription& device_info) {
+  return nvptx::LibDevicePath(
+      hlo_config.debug_options().xla_gpu_cuda_data_dir());
+}
+
+}  // namespace gpu
+}  // namespace xla
```
Comment (ddunl) [inline xla/service/gpu/ir_emitter_triton_cuda.cc:126]: Could you add a newline at the end of this file (also `ir_emitter_triton_rocm.cc`)

Comment (zoranjovanovic-ns) [reply xla/service/gpu/ir_emitter_triton_cuda.cc:126]: Yes, I will do that.

**Review by ddunl — APPROVED** [review]
Thanks!!

**Comment by i-chaochen** [comment]
Hi @xla-rotation  this PR is already in the copybara-service pipeline, but it has conflicts now. Shall we resolve and rebase it? or we can leave it?

**Comment by ddunl** [comment]
I think you can leave it, we have to patch because Copybara behaved strangely on this one so I have to investigate that, fix and patch

**Comment by zoranjovanovic-ns** [comment]
> I think you can leave it, we have to patch because Copybara behaved strangely on this one so we have to investigate that, fix and patch

Hi @xla-rotation, should I take some additional actions in order to have this merged?

**Comment by penpornk** [comment]
@zoranjovanovic-ns Sorry for the delay! Could you please help resolve merge conflicts?

**Comment by zoranjovanovic-ns** [comment]
Yes, I will rebase and push new version.

**Comment by penpornk** [comment]
@zoranjovanovic-ns Thank you for checking and for merging conflicts! We are applying some more fixes to pass internal tests. I'll let you know if we need more help. :)

**Pull request closed without merging**